app/Jobs/RemoveBeatmapsetScores.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+declare(strict_types=1);
+
+namespace App\Jobs;
+
+use App\Libraries\Search\ScoreSearch;
+use App\Libraries\Search\ScoreSearchParams;
+use App\Models\Beatmap;
+use App\Models\Beatmapset;
+use App\Models\Score\Best\Model as ScoreBestModel;
+use App\Models\Solo\Score;
+use App\Models\UserStatistics;
+use Ds\Set;
+use Illuminate\Bus\Queueable;
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Database\Eloquent\Collection;
+
+class RemoveBeatmapsetScores implements ShouldQueue
+{
+    use Queueable;
+
+    public $timeout = 3600;
+
+    private int $beatmapsetId;
+    private int $maxScoreId;
+    private Set $resetUserRankStats;
+    private array $schemas;
+    private ScoreSearch $scoreSearch;
+
+    /**
+     * Create a new job instance.
+     *
+     * @return void
+     */
+    public function __construct(Beatmapset $beatmapset)
+    {
+        $this->beatmapsetId = $beatmapset->getKey();
+        $this->maxScoreId = Score::max('id') ?? 0;
+    }
+
+    /**
+     * Execute the job.
+     *
+     * @return void
+     */
+    public function handle()
+    {
+        $this->resetUserRankStats = new Set();
+        $this->scoreSearch = new ScoreSearch();
+        $this->schemas = $this->scoreSearch->getActiveSchemas();
+
+        Score::with('performance')->whereIn('beatmap_id', Beatmap::where('beatmapset_id', $this->beatmapsetId)->select('beatmap_id'))->where('id', '<=', $this->maxScoreId)->chunkById(1000, function (Collection $scores): void {
+            foreach ($scores as $score) {
+                $this->deleteScore($score);
+            }
+        });
+    }
+
+    private function deleteScore(Score $score): void
+    {
+        $this->resetUserRankStatsFor($score);
+        $score->delete();
+        $score->performance?->delete();
+        $this->scoreSearch->queueForIndex($this->schemas, [$score->getKey()]);
```

Comment:
Can do with a transaction, otherwise errors can cause the score to be already deleted from the database but not be queued yet; they won't be picked up in `handle` if the job is run again since they're already gone from the database. (or delete the score from index first like the other job)

Reply:
Also, this is a lot of delete queries 👀 
`RemoveBeatmapsetBestScores` triggered `delete` on each score to run the extra delete logic on them, this doesn't do any yet...

Also not sure if the default job timeout is going to be an issue with all the delete queries and es requests in `resetUserRankStatsFor`; all the jobs during my testing with 10k scores per beatmap have all managed to exceed the default 60s timeout and failed.

Reply:
timeout should be 3600 as set in the class 🤔  are you observing that it's not the case?

also the `resetUserRankStatsFor` kind of made it difficult to have complete `delete` function so I figured everything should just go here instead for now.

Reply:
It seems that `queue:listen` has its own timeout that overrides the job's timeout but `queue:work` prioritizes the correct one psy🦆